File: y11592/src/services/QueueWorker.ts
import { CronJob } from 'cron';
import { config } from '../config';
import logger from '../utils/logger';
import { retryQueueService } from './RetryQueueService';

class QueueWorker {
  private job: CronJob | null = null;
  private isRunning: boolean = false;

  start(): void {
    if (this.job) {
      logger.warn('队列工作器已在运行');
      return;
    }

    const cronExpression = `*/${config.queue.workerIntervalSeconds} * * * * *`;
    
    this.job = new CronJob(
      cronExpression,
      async () => {
        if (this.isRunning) {
          logger.debug('跳过执行，上一批次仍在处理中');
          return;
        }

        this.isRunning = true;
        try {
          const result = await retryQueueService.processPendingItems();
          if (result.processed > 0) {
            logger.info(`队列处理完成: 处理 ${result.processed} 条, 成功 ${result.success} 条, 失败 ${result.failed} 条`);
          }
        } catch (error) {
          logger.error('队列处理异常:', error);
        } finally {
          this.isRunning = false;
        }
      },
      null,
      true,
      'Asia/Shanghai'
    );

    logger.info(`队列工作器已启动，每 ${config.queue.workerIntervalSeconds} 秒执行一次`);
  }

  stop(): void {
    if (this.job) {
      this.job.stop();
      this.job = null;
      logger.info('队列工作器已停止');
    }
  }

  triggerManually(): Promise<{ processed: number; success: number; failed: number }> {
    logger.info('手动触发队列处理');
    return retryQueueService.processPendingItems();
  }
}

export const queueWorker = new QueueWorker();

File: y11577/app/services/queue_worker.py
import logging
from datetime import datetime
from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.interval import IntervalTrigger

from app.config import settings
from app.database import SessionLocal
from app.services.queue_service import CompensationQueueService, QueueStatus

logger = logging.getLogger(__name__)


class QueueWorker:
    def __init__(self):
        self.scheduler = BackgroundScheduler()
        self.is_running = False

    def start(self):
        if self.is_running:
            logger.warning("Queue worker is already running")
            return
        
        if not settings.QUEUE_WORKER_ENABLED:
            logger.info("Queue worker is disabled by configuration")
            return
        
        trigger = IntervalTrigger(seconds=settings.QUEUE_WORKER_INTERVAL_SECONDS)
        self.scheduler.add_job(
            self._process_queue,
            trigger=trigger,
            id="queue_processing",
            name="队列自动处理",
            replace_existing=True
        )
        
        self.scheduler.start()
        self.is_running = True
        logger.info(
            "Queue worker started, processing every %d seconds",
            settings.QUEUE_WORKER_INTERVAL_SECONDS
        )

    def stop(self):
        if self.is_running:
            self.scheduler.shutdown(wait=False)
            self.is_running = False
            logger.info("Queue worker stopped")

    def _process_queue(self):
        db = SessionLocal()
        try:
            queue_service = CompensationQueueService(db)
            pending_items = queue_service.get_pending_items(batch_size=10)
            
            if not pending_items:
                logger.debug("No pending queue items to process")
                return
            
            logger.info("Processing %d pending queue items", len(pending_items))
            
            for item in pending_items:
                try:
                    result = queue_service.process_item(item.id)
                    if result["success"]:
                        logger.info(
                            "Queue item %d processed successfully (type: %s)",
                            item.id,
                            item.business_type
                        )
                    else:
                        logger.warning(
                            "Queue item %d processing failed: %s (type: %s, retry: %d/%d)",
                            item.id,
                            result.get("error", "Unknown"),
                            item.business_type,
                            item.retry_count,
                            item.max_retries
                        )
                except Exception as e:
                    logger.error(
                        "Exception processing queue item %d: %s",
                        item.id,
                        str(e),
                        exc_info=True
                    )
                    
        except Exception as e:
            logger.error("Error in queue worker: %s", str(e), exc_info=True)
        finally:
            db.close()


queue_worker = QueueWorker()
